repo/repo.go

```diff
@@ -98,33 +91,95 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
 			if err != nil {
 				return nil, nil, err
 			}
+			leveled := strings.HasPrefix(f, "+")
+			f = strings.TrimPrefix(f, "+")
 			f = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(f))
-			macroFile, err := rule.LoadMacroFile(f, "", defName)
-			if err != nil {
+			checked[f+"%"+defName] = true
+
+			la := &loadArgs{
+				leveled: leveled,
+				workspace: workspace.Path,
+				f: f,
+				defName: defName,
+				repos: repos,
+				repoFileMap: repoFileMap,
+				repoIndexMap: repoIndexMap,
+				checked: checked,
+			}
+
+			if err := loadRepositoriesFromMacro(la); err != nil {
 				return nil, nil, err
 			}
-			for _, repo := range macroFile.Rules {
-				if name := repo.Name(); name != "" {
-					repos = append(repos, repo)
-					repoFileMap[name] = macroFile
-					repoIndexMap[name] = len(repos) - 1
+			repos = la.repos
+		}
+	}
+	return repos, repoFileMap, nil
+}
+
+type loadArgs struct {
+	leveled bool
+	workspace string
+	f string
+	defName string
+	repos []*rule.Rule
+	repoFileMap map[string]*rule.File
+	repoIndexMap map[string]int
+	checked map[string]bool
+}
+
+func loadRepositoriesFromMacro(la *loadArgs) error {
+	macroFile, err := rule.LoadMacroFile(la.f, "", la.defName)
+	if err != nil {
+		return err
+	}
+	for _, repo := range macroFile.Rules {
+		if name := repo.Name(); name != "" {
+				la.repos = append(la.repos, repo)
+				la.repoFileMap[name] = macroFile
+				la.repoIndexMap[name] = len(la.repos) - 1
+		} else if la.leveled {
+			// This block will only be entered if leveled==true
+			var callFile string
+			kind := repo.Kind()
+			for _, l := range macroFile.Loads {
+				if l.Has(kind) {
+					pkg := filepath.Join(filepath.Dir(la.workspace), filepath.Clean(l.Name()))
+					callFile = strings.Replace(pkg, ":", string(filepath.Separator), 1)
+					la.defName = l.Unalias(kind)
+					break
 				}
 			}
-			extraRepos, err = parseRepositoryDirectives(macroFile.Directives)
-			if err != nil {
-				return nil, nil, err
+			// TODO: Also handle the case where one macro calls another macro in the same bzl file
+			if len(callFile) == 0 {
+				continue
 			}
-			for _, repo := range extraRepos {
-				if i, ok := repoIndexMap[repo.Name()]; ok {
-					repos[i] = repo
-				} else {
-					repos = append(repos, repo)
+			if !la.checked[la.f+"%"+la.defName] {
+				la.f = callFile
+				la.checked[la.f+"%"+la.defName] = true
+				if err := loadRepositoriesFromMacro(la); err != nil {
+					return err
 				}
-				repoFileMap[repo.Name()] = macroFile
 			}
 		}
 	}
-	return repos, repoFileMap, nil
+	la.repos, err = loadExtraRepos(macroFile, la.repos, la.repoFileMap, la.repoIndexMap)
```

Comment:
This is for backwards compatibility: Gazelle still operates normally with directives declared using a `# gazelle:repository` macro instead of a bazel rule, so I simply just wanted to still allow this. For example, see repositories declared in https://github.com/bazelbuild/bazel-gazelle/blob/master/WORKSPACE

Reply:
I see. gazelle:repository directive was already supported in macros before this change